clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -204,9 +206,19 @@ public double compressionRate() {
     public boolean isFull() {
         return recordsBuilder.isFull();
     }
-
-    public void close() {
-        recordsBuilder.close();
+    /**
+     * Writes the final metadata in the batch and updates transaction state with the new sequence number for the
+     * TopicPartition
+     * @param transactionState the global transaction state which keeps track of sequence numbers for each TopicPartition.
+     */
+    public synchronized void close(TransactionState transactionState) {
+        if (!isClosed) {
+            recordsBuilder.close();
+            if (transactionState != null) {
+                transactionState.incrementSequenceNumber(topicPartition, this.recordCount);
```

Comment:
Sorry to belabor this, but if we instead expose `recordCount` as a method, then we could let the caller do this. For example, from `RecordAccumulator`, we could do this:

```java
batch.close();
if (transactionState != null)
  transactionState.incrementSequenceNumber(batch.partition(), batch.recordCount());
```
It may not be a huge advantage, but it keeps these classes decoupled.